src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts

```diff
@@ -165,31 +146,72 @@ export class WorkspaceSavedObjectsClientWrapper {
       id: string,
       options: SavedObjectsDeleteOptions = {}
     ) => {
-      if (this.isRelatedToWorkspace(type)) {
-        await this.validateSingleWorkspacePermissions(id, wrapperOptions.request, [
-          WorkspacePermissionMode.Management,
-        ]);
-      }
-
       const objectToDeleted = await wrapperOptions.client.get(type, id, options);
-      await this.validateMultiWorkspacesPermissions(
-        objectToDeleted.workspaces,
+      const workspacePermitted = await this.validateMultiWorkspacesPermissions(
+        objectToDeleted.workspaces!,
         wrapperOptions.request,
         [WorkspacePermissionMode.LibraryWrite, WorkspacePermissionMode.Management]
       );
+
+      if (!workspacePermitted) {
+        const objectsPermitted = await this.validateObjectsPermissions(
+          [{ type, id }],
+          wrapperOptions.request,
+          [
+            WorkspacePermissionMode.Management,
+            WorkspacePermissionMode.LibraryWrite,
+            WorkspacePermissionMode.Write,
+          ]
+        );
+        if (!objectsPermitted) {
+          throw generateSavedObjectsPermissionError();
+        }
+      }
       return await wrapperOptions.client.delete(type, id, options);
     };
 
+    // validate single object update with workspace permission, which is used for update and bulkUpdate
+    const validateUpdateWithWorkspacePermission = async <T = unknown>(
+      type: string,
+      id: string,
+      options: SavedObjectsUpdateOptions = {}
+    ): Promise<boolean> => {
+      const objectToUpdate = await wrapperOptions.client.get<T>(type, id, options);
+      let workspacePermitted = false;
+      if (objectToUpdate.workspaces && objectToUpdate.workspaces.length > 0) {
+        workspacePermitted =
+          (await this.validateMultiWorkspacesPermissions(
```

Comment:
Should it call `validateAtLeastOnePermittedWorkspaces` here? I remember a user can update the object if the user has permission to any of the workspace that the object is associated with, am I right? @wanglam @SuZhou-Joe

Reply:
Yes, I agree with @ruanyl that here should be AtLeastOne